### PR TITLE
Add uv_replace_allocator to allow custom malloc/free.

### PR DESCRIPTION
**Pull request by beevik**
Modified uv.h to add a new api: uv_replace_allocator.
Modified uv-common.h and uv-common.cpp to add new internal uv__malloc & uv_free APIs.
Modified all code in src to use uv__malloc/uv__free instead of malloc/free.
Modified documentation file misc.rst to describe how uv_replace_allocator works.

### CODE REVIEW

**Comment by beevik** [comment]
I replaced my previous PR with this one, for a couple reasons:
* My old PR was based on v1.x instead of master. I shouldn't have based off of v1.x since this is a new feature.
* This cleans up the commit history.

If you agree that this is a good idea, I'll drop my old PR.

File: src/unix/fsevents.c [hunk 1]
```diff
@@ -880,7 +880,7 @@ int uv__fsevents_close(uv_fs_event_t* handle) {
   /* Wait for deinitialization */
   uv_sem_wait(&state->fsevent_sem);
 
-  uv_close((uv_handle_t*) handle->cf_cb, (uv_close_cb) free);
+  uv_close((uv_handle_t*) handle->cf_cb, (uv_close_cb) uv__free);
```
Comment (beevik) [inline src/unix/fsevents.c]: I'd like some extra eyes on this change, since this is an unusual use of free.

File: src/uv-common.c [hunk 2]
```diff
@@ -33,6 +33,33 @@
...
+
+void uv__free(void* ptr) {
+  if (replaced_free)
+    (*replaced_free)(ptr);
+  else
+    free(ptr);
+}
+
+
+void uv_replace_allocator(uv_malloc_func malloc_func, uv_free_func free_func) {
```
Comment (saghul) [inline src/uv-common.c]: Release builds can still shoot themselves in the foot. I'd go for making this return an int, and if the allocator was already set we return UV_EINVAL for example.

**Comment by saghul** [comment]
We do allow new features in v1.x, as long as they are backwards API compatible and they don't break the ABI.  This PR fulfills both requirements, so I'd be willing to add it to v1.x.

/cc @bnoordhuis @piscisaureus

**Comment by beevik** [comment]
Based on @sagul's comment, I'm closing this and will continue to make changes in the v1.x branch.

**Pull request closed without merging**